benchmarks/helics/PholdFederate.hpp

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright (c) 2017-2019,
+Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance for Sustainable Energy, LLC.  See
+the top-level NOTICE for additional details. All rights reserved.
+SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#include "helics/application_api/Endpoints.hpp"
+#include "helics/application_api/MessageFederate.hpp"
+#include "helics/core/ActionMessage.hpp"
+#include "helics/helics-config.h"
+#include <random>
+
+using namespace helics;
+/** class implementing a federate for the PHOLD benchmark*/
+class PholdFederate
+{
+  public:
+    helics::Time deltaTime = helics::Time (10, time_units::ns);  // sampling rate
+    helics::Time finalTime = helics::Time (10000, time_units::ns);  // final time
+    int evCount = 0; // number of events handled by this federate
+
+  private:
+    std::unique_ptr<helics::MessageFederate> mFed;
+    helics::Endpoint* ept;
+
+    int index_ = 0;
+    int maxIndex_ = 0;
+    int initEvCount_ = 16; // starting number of events
+    double localProbability_ = .1; // probability of local events
+    double randTimeMean_ = deltaTime * 2; // mean for the exponential distribution used when picking event times
+
+    // classes related to the exponential distribution random number generator
+    std::random_device rd;
+    std::mt19937 rand_gen;
+    std::exponential_distribution<> rand_exp;
+
+    bool initialized{false};
+    bool readyToRun{false};
+
+  public:
+    PholdFederate () = default;
+
+    void run (std::function<void ()> callOnReady = nullptr)
+    {
+        makeReady ();
+        if (callOnReady)
+        {
+            callOnReady ();
+        }
+        mainLoop ();
+    };
+
+    void initialize (const std::string &coreName, int index, int maxIndex)
+    {
+        std::string name = "phold_" + std::to_string (index);
+        index_ = index;
+        maxIndex_ = maxIndex;
+        helics::FederateInfo fi;
+        fi.coreName = coreName;
+        mFed = std::make_unique<helics::MessageFederate> (name, fi);
+        ept = &mFed->registerEndpoint ("ept");
+
+        // en.cppreference.com/w/cpp/numeric/random/exponential_distribution
+        rand_gen.seed(rd());
+        rand_exp = std::exponential_distribution<> (1/randTimeMean_);
```

Comment:
might need "1.0" instead so some compiler doesn't get the idea to do integer division and just to make it clear this is a floating point value